app/controllers/single_pages_controller.rb

```diff
@@ -122,8 +123,201 @@ def export_to_excel
     end
   end
 
+  def upload_samples
```

Comment:
this is a hell of a long method that would benefit and be easier to understand if split up, or made use of some classes